src/base/materialized.ts

```diff
@@ -230,84 +231,148 @@ export class MaterializedChatMessage implements IMaterializedNode {
 		return tokenizer.countMessageTokens({ ...this.toChatMessage(), content: '' });
 	});
 
-	private readonly _text = once(() => {
-		let result = '';
+	private readonly _text = once((): (string | MaterializedChatMessageImage)[] => {
+		let result: (string | MaterializedChatMessageImage)[] = [];
 		for (const { text, isTextSibling } of textChunks(this)) {
+			if (text instanceof MaterializedChatMessageImage) {
+				result.push(text);
+				continue;
+			}
 			if (
 				text.lineBreakBefore === LineBreakBefore.Always ||
 				(text.lineBreakBefore === LineBreakBefore.IfNotTextSibling && !isTextSibling)
 			) {
-				if (result.length && !result.endsWith('\n')) {
-					result += '\n';
+				let prev = result[result.length - 1]
+				if (typeof prev === 'string' && result.length && !prev.endsWith('\n')) {
+					result.push('\n');
```

Comment:
```suggestion
				if (typeof prev === 'string' && !prev.endsWith('\n')) {
					result[result.length - 1] = prev + '\n';
```

the length check is redundant, and we should avoid making extra string parts